### PR TITLE
[DPP+1327] Add ability to set up a hierarchy of concurrency limiters [ON-HOLD]

### PR DESCRIPTION
**Pull request by pbatko-da**
Extracted from https://github.com/digital-asset/daml/pull/15462

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 1]
```diff
@@ -14,20 +14,27 @@ class NoConcurrencyLimiter extends ConcurrencyLimiter {
   override def execute[T](task: => Future[T]): Future[T] = task
 }
 
+/** @param executionContext - needed to run limiter's internal book-keeping. Does not run the submitted tasks.
+  * @param parentO - used to define a hierarchy (a tree) of concurrency limiters such that we can
+  *                  express constraints like: for 5 local limiters each running at most 3 concurrent tasks
+  *                  set a global limit of at most 10 concurrent in total.
+  */
+// TODO etq: Separate 1) concurrency limiting from 2) composition of limiters
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:22]: Isn't this is the same as to have two original ConcurrencyLimiter used so: child.execute(parent.execute(stuff))?
If yes, I would suggest not changing this file.

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 2]
```diff
@@ -49,7 +56,15 @@ class QueueBasedConcurrencyLimiter(
     while (running < parallelism && waiting.nonEmpty) {
       val head = waiting.dequeue()
       running = running + 1
-      head()
+      parentO match {
+        case Some(parent) =>
+          // defering creating the tasks's future to the parent concurrency limiter
+          parent.execute(head())
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:62]: this is the right ordering. blocked parent starves the child pool ("running" futures doing nothing). The other way around would be bad.

Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:62]: in practice, without these changes it should go like: locallimitter.execute(globallimitter.execute(stuff))

Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:62]: Is your recommendation to cancel the changes from this PR and instead compose then at call sites like in your example?

Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:62]: I could see value in defining some combination function on the companion for the trait, which does the right order in composing limiters, and adds tests for them, that 
* global limits are respected
* global limiter is not flooded by local limiting waiting tasks

**Pull request closed without merging**